File: golang/2202/220218.go
package main

import (
	"fmt"
)

func removeKdigits(num string, k int) (ret string) {
	stack := make([]int32, 0)
	for _, v := range num {
		for k > 0 && len(stack) > 0 && stack[len(stack)-1] > v {
			stack = stack[:len(stack)-1]
			k--
		}
		stack = append(stack, v)
	}
	if k > 0 {
		stack = stack[:len(stack)-k]
	}
	for _, v := range stack {
		ret += string(v)
	}
	for len(ret) > 1 && ret[0] == '0' {
		ret = ret[1:]
	}
	if len(ret) == 0 {
		return "0"
	}
	return
}

func main() {
	fmt.Println(removeKdigits("1432219", 3))
}

File: cpp/2202/220218.cpp
#include <string>
#include <vector>
#include <iostream>

/**
 * 402. Remove K Digits
 * Given string num representing a non-negative integer num, and an integer k, return the smallest possible integer after removing k digits from num.
 */

class Solution {
public:
	static std::string removeKdigits(const std::string& num, int k) {
		std::vector<char> stack;
		stack.reserve(num.length() - k);
		for (char ch : num) {
			while (k && !stack.empty() && stack.back() > ch) {
				stack.pop_back();
				--k;
			}
			stack.push_back(ch);
		}
		std::string ret(stack.begin(), k > 0 ? std::prev(stack.end(), k) : stack.end());
		return ret.empty() || ret.find_first_not_of('0') == -1 ? "0" : ret.substr(ret.find_first_not_of('0'));
	}
};

int main() {
	std::cout << Solution::removeKdigits("1432219", 3);
	return 0;
}
